cmd/cloudinit/cloudinit.go

```diff
@@ -204,49 +124,67 @@ func saveCloudConfig() error {
 		scriptBytes = userDataBytes
 		userDataBytes = []byte{}
 	} else if isCompose(userData) {
-		if userDataBytes, err = toCompose(userDataBytes); err != nil {
-			log.Errorf("Failed to convert to compose syntax: %v", err)
+		if userDataBytes, err = composeToCloudConfig(userDataBytes); err != nil {
+			log.Errorf("Failed to convert compose to cloud-config syntax: %v", err)
 			return err
 		}
 	} else if config.IsCloudConfig(userData) {
-		if rancherConfig.ReadConfig(userDataBytes) == nil {
+		if rancherConfig.ReadConfig(userDataBytes, false) == nil {
 			log.WithFields(log.Fields{"cloud-config": userData}).Warn("Failed to parse cloud-config, not saving.")
 			userDataBytes = []byte{}
 		}
 	} else {
-		log.Errorf("Unrecognized cloud-init\n%s", userData)
+		log.Errorf("Unrecognized user-data\n%s", userData)
 		userDataBytes = []byte{}
 	}
 
-	userDataBytesMerged, scriptBytes, err := mergeBaseConfig(userDataBytes, scriptBytes)
-	if err != nil {
-		log.Errorf("Failed to merge base config: %v", err)
-	} else if rancherConfig.ReadConfig(userDataBytesMerged) == nil {
-		log.WithFields(log.Fields{"cloud-config": userData}).Warn("Failed to parse merged cloud-config, not merging.")
-	} else {
-		userDataBytes = userDataBytesMerged
+	if rancherConfig.ReadConfig(userDataBytes, false) == nil {
+		log.WithFields(log.Fields{"cloud-config": userData}).Warn("Failed to parse cloud-config")
+		return errors.New("Failed to parse cloud-config")
 	}
 
 	return saveFiles(userDataBytes, scriptBytes, metadata)
 }
 
+func fetchUserData() ([]byte, datasource.Metadata, error) {
+	var metadata datasource.Metadata
+	ds, err := currentDatasource()
+	if err != nil || ds == nil {
+		log.Errorf("Failed to select datasource: %v", err)
+		return nil, metadata, err
+	}
+	log.Infof("Fetching user-data from datasource %v", ds.Type())
+	userDataBytes, err := ds.FetchUserdata()
+	if err != nil {
+		log.Errorf("Failed fetching user-data from datasource: %v", err)
+		return nil, metadata, err
+	}
+	log.Infof("Fetching meta-data from datasource of type %v", ds.Type())
+	metadata, err = ds.FetchMetadata()
+	if err != nil {
+		log.Errorf("Failed fetching meta-data from datasource: %v", err)
+		return nil, metadata, err
+	}
+	return userDataBytes, metadata, nil
+}
+
 func getSaveCloudConfig() (*config.CloudConfig, error) {
-	ds := file.NewDatasource(rancherConfig.CloudConfigFile)
+	ds := file.NewDatasource(rancherConfig.CloudConfigBootFile)
```

Comment:
I don't think this works right.  Before, a bit back, we used to merge everything to one file called cloud-config-processed.yml and that was read.  Now I believe we need to just load the default RancherOS config and used that as our cloud-config.  We don't need data sources at all now because we already have the data structures saved to disk.